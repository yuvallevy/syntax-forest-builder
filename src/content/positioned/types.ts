import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type Width = number;

export type PlotCoords = {
  plotX: PlotX;
  plotY: PlotY;
};

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: Id[];
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  slice: StringSlice;
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedStrandedNode = NodeCommon & WithPositionInTree;

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode | PositionedStrandedNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: PlotCoords;
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
