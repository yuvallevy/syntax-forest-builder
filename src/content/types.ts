export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;

export type StringSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type NodeIndicatorInPlot = {
  treeId: Id;
  nodeId: Id;
};

export type TreeCommon = {
  sentence: Sentence;
}

export type NodeCommon = {
  label: NodeLabel;
};
