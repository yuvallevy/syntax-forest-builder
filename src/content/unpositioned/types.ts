import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;

export type PlotCoordsOffset = {
  dPlotX: DPlotX;
  dPlotY: DPlotY;
};

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: PlotCoordsOffset;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: Id[];
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: StringSlice;
  triangle: boolean;
};

export type UnpositionedStrandedNode = NodeCommon & WithOffsetInTree & {
  formerDescendants?: IdMap<UnpositionedNode>;
  formerSlice?: StringSlice;
  formerlyTriangle?: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode | UnpositionedStrandedNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export const isTerminal = (node: UnpositionedNode): node is UnpositionedTerminalNode =>
  'slice' in node;
