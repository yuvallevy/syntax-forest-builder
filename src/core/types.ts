export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;
type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;
type Width = number;

export type NodeSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type UnpositionedForest = { plots: IdMap<UnpositionedPlot>; };

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

type TreeCommon = {
  sentence: Sentence;
}

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: { dPlotX: DPlotX; dPlotY: DPlotY; };
};

export type NodeCommon = {
  label: NodeLabel;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: IdMap<UnpositionedNode>;
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: NodeSlice;
  triangle: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: IdMap<PositionedNode>;
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: {
    plotX: PlotX;
    plotY: PlotY;
  };
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
