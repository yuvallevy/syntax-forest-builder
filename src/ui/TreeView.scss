.TreeView-node {
  cursor: pointer;
}

.TreeView-node-selected {
  fill: cornflowerblue;
}
