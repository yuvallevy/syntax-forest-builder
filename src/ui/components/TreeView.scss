.TreeView-node {
  cursor: pointer;
}

.TreeView-node rect {
  fill: transparent;
  stroke: none;
}

.TreeView-node:hover rect {
  fill: #eee;
}

.TreeView-node-empty-label text {
  fill: #888;
}

.TreeView-node-selected rect {
  fill: #fafafa;
  stroke: #444;
  stroke-dasharray: 4 4;
  stroke-width: 1;
  animation: marching-ants .75s linear infinite;
}

.TreeView-node-ghost {
  fill: none;
  stroke: #444;
  stroke-dasharray: 4 4;
  stroke-width: 1;
  opacity: 0.6;
}

.NodeCreationTriggerClickZone-area {
  cursor: pointer;
  stroke: none;
  fill: transparent;
}

.NodeCreationTriggerClickZone-indicator {
  stroke: #999;
  fill: none;
  pointer-events: none;
  display: none;
}

.NodeCreationTriggerClickZone:hover .NodeCreationTriggerClickZone-indicator {
  display: unset;
}

@keyframes marching-ants {
  0% {
    stroke-dashoffset: 0;
  }
  100% {
    stroke-dashoffset: -8px;
  }
}
