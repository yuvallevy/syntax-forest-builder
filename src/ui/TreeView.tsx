import React from 'react';
import { mapEntries } from '../core/objTransforms';
import { Id, PositionedBranchingNode, PositionedNode, PositionedTerminalNode, PositionedTree } from '../core/types';
import './TreeView.scss';

const NODE_LEVEL_SPACING = 20;
const TRIANGLE_BASE_Y = -2;

interface TreeViewProps {
  treeId: Id;
  tree: PositionedTree;
  selectedNodeIds: Id[];
  onNodeSelect?: (nodeId: Id) => void;
}

const renderChildNodeConnections = (node: PositionedBranchingNode): React.ReactNode[] =>
  mapEntries(node.children, ([childId, child]) =>
    <line
      key={`to-${childId}`}
      stroke="#000"
      x1={node.position.treeX}
      y1={node.position.treeY}
      x2={child.position.treeX}
      y2={child.position.treeY - NODE_LEVEL_SPACING}
    />
  );

const renderTriangleConnection = (nodeId: Id, node: PositionedTerminalNode): React.ReactNode =>
  node.triangle && <path
    key={`triangle-${nodeId}`}
    stroke="#000"
    fill="none"
    d={`M${node.position.treeX} ${node.position.treeY} L${node.triangle.treeX1} ${TRIANGLE_BASE_Y}  L${node.triangle.treeX2} ${TRIANGLE_BASE_Y} Z`}
  />;

const renderNodeFlat =
  (nodeId: Id, node: PositionedNode, selectedNodeIds: Id[], onSelect?: (id: Id) => void): React.ReactNode[] => [
  <text
    key={nodeId}
    x={node.position.treeX}
    y={node.position.treeY}
    className={'TreeView-node' + (selectedNodeIds.includes(nodeId) ? ' TreeView-node-selected' : '')}
    fill="#000"
    textAnchor="middle"
    dominantBaseline="text-after-edge"
    onMouseDown={() => onSelect && onSelect(nodeId)}
  >
    {node.label}
  </text>,
  'triangle' in node && renderTriangleConnection(nodeId, node),
  ...('children' in node)
    ? [
      ...renderChildNodeConnections(node),
      ...mapEntries(node.children, ([nodeId, node]) => renderNodeFlat(nodeId, node, selectedNodeIds, onSelect)),
    ] : [],
];

const TreeView: React.FC<TreeViewProps> = ({ treeId, tree, selectedNodeIds, onNodeSelect }) =>
  <g id={`tree-${treeId}`} style={{ transform: `translate(${tree.position.plotX}px, ${tree.position.plotY}px)` }}>
    {mapEntries(tree.nodes, ([nodeId, node]) => renderNodeFlat(nodeId, node, selectedNodeIds, onNodeSelect))}
  </g>;

export default TreeView;
