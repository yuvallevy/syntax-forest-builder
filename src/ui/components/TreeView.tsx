import React from 'react';
import { mapEntries } from '../../util/objTransforms';
import {
  Id, IdMap
} from '../../content/types';
import './TreeView.scss';
import { NodeSelectionMode } from '../selection';
import { NodeCreationTrigger, getNodeCreationTriggersForTree } from '../nodeCreationTriggers';
import { ClientCoordsOffset } from '../coords';
import strWidth from '../strWidth';
import {
  PositionedBranchingNode, PositionedNode, PositionedTerminalNode, PositionedTree
} from '../../content/positioned/types';

const NODE_LEVEL_SPACING = 20;
const TRIANGLE_BASE_Y = -2;

const NODE_AREA_WIDTH = 35;
const NODE_AREA_HEIGHT = 20;
const NODE_AREA_RELATIVE_X = -(NODE_AREA_WIDTH / 2);
const NODE_AREA_RELATIVE_Y = -18.5;

interface NodeCreationTriggerClickZoneProps {
  trigger: NodeCreationTrigger;
  onClick?: () => void;
}

interface TreeViewProps {
  treeId: Id;
  tree: PositionedTree;
  selectedNodeIds: Id[];
  nodeDragOffset?: ClientCoordsOffset;
  onNodeMouseDown?: (event: React.MouseEvent<SVGElement>) => void;
  onSingleNodeSelect?: (nodeId: Id, mode: NodeSelectionMode) => void;
  onNodeCreationTriggerClick?: (trigger: NodeCreationTrigger) => void;
}

const renderChildNodeConnections = (node: PositionedBranchingNode, allNodes: IdMap<PositionedNode>): React.ReactNode[] =>
  node.children.map(childId =>
    <line
      key={`to-${childId}`}
      stroke="#000"
      x1={node.position.treeX}
      y1={node.position.treeY}
      x2={allNodes[childId].position.treeX}
      y2={allNodes[childId].position.treeY - NODE_LEVEL_SPACING}
    />
  );

const renderTriangleConnection = (nodeId: Id, node: PositionedTerminalNode): React.ReactNode =>
  node.triangle && <path
    key={`triangle-${nodeId}`}
    stroke="#000"
    fill="none"
    d={`M${node.position.treeX} ${node.position.treeY} L${node.triangle.treeX1} ${TRIANGLE_BASE_Y}  L${node.triangle.treeX2} ${TRIANGLE_BASE_Y} Z`}
  />;

const renderNode = (
  nodeId: Id,
  node: PositionedNode,
  allNodes: IdMap<PositionedNode>,
  selectedNodeIds: Id[],
  nodeDragOffset?: ClientCoordsOffset,
  onMouseDown?: (event: React.MouseEvent<SVGElement>) => void,
  onSelect?: (id: Id, mode: NodeSelectionMode) => void,
): React.ReactNode[] => [
  <g
    key={nodeId}
    className={'TreeView-node' + (selectedNodeIds.includes(nodeId) ? ' TreeView-node-selected' : '')}
    onMouseDown={event => {
      onSelect && onSelect(nodeId, event.ctrlKey || event.metaKey ? 'ADD' : 'SET');
      onMouseDown && onMouseDown(event);
    }}
  >
    <rect
      x={node.position.treeX + NODE_AREA_RELATIVE_X}
      y={node.position.treeY + NODE_AREA_RELATIVE_Y}
      width={NODE_AREA_WIDTH}
      height={NODE_AREA_HEIGHT}
      rx={3}
      ry={3}
    />
    <text
      x={node.position.treeX}
      y={node.position.treeY}
      fill="#000"
      textAnchor="middle"
      dominantBaseline="text-after-edge"
    >
      {node.label}
    </text>
  </g>,
  nodeDragOffset && selectedNodeIds.includes(nodeId) && <rect
    key={`${nodeId}-ghost`}
    className="TreeView-node-ghost"
    x={node.position.treeX + NODE_AREA_RELATIVE_X + nodeDragOffset.dClientX}
    y={node.position.treeY + NODE_AREA_RELATIVE_Y + nodeDragOffset.dClientY}
    width={NODE_AREA_WIDTH}
    height={NODE_AREA_HEIGHT}
    rx={3}
    ry={3}
  />,
  'triangle' in node && renderTriangleConnection(nodeId, node),
  'children' in node && renderChildNodeConnections(node, allNodes),
];

const NodeCreationTriggerClickZone: React.FC<NodeCreationTriggerClickZoneProps> = ({ trigger, onClick }) =>
  <g
    className="NodeCreationTriggerClickZone"
    onClick={onClick}
  >
    <rect
      className="NodeCreationTriggerClickZone-area"
      x={trigger.topLeft.treeX}
      y={trigger.topLeft.treeY}
      width={trigger.bottomRight.treeX - trigger.topLeft.treeX}
      height={trigger.bottomRight.treeY - trigger.topLeft.treeY}
    />
    <circle
      className="NodeCreationTriggerClickZone-indicator"
      cx={trigger.origin.treeX}
      cy={trigger.origin.treeY}
      r={8}
    />
    {'childIds' in trigger
      ? trigger.childPositions.map(childPosition => <line
        key={`${childPosition.treeX},${childPosition.treeY}`}
        className="NodeCreationTriggerClickZone-indicator"
        x1={trigger.origin.treeX}
        y1={trigger.origin.treeY}
        x2={childPosition.treeX}
        y2={childPosition.treeY - NODE_LEVEL_SPACING}
      />)
      : <line
        className="NodeCreationTriggerClickZone-indicator"
        x1={trigger.origin.treeX}
        y1={trigger.origin.treeY}
        x2={trigger.origin.treeX}
        y2={0}
      />}
  </g>;

const TreeView: React.FC<TreeViewProps> = ({
  treeId,
  tree,
  selectedNodeIds,
  nodeDragOffset,
  onNodeMouseDown,
  onSingleNodeSelect,
  onNodeCreationTriggerClick,
}) =>
  <g id={`tree-${treeId}`} style={{ transform: `translate(${tree.position.plotX}px, ${tree.position.plotY}px)` }}>
    {getNodeCreationTriggersForTree(strWidth)(tree).map(trigger => <NodeCreationTriggerClickZone
      trigger={trigger}
      key={'childIds' in trigger ? trigger.childIds.join() : trigger.slice.join()}
      onClick={() => onNodeCreationTriggerClick && onNodeCreationTriggerClick(trigger)}
    />)}
    {mapEntries(tree.nodes, ([nodeId, node]) =>
      renderNode(nodeId, node, tree.nodes, selectedNodeIds, nodeDragOffset, onNodeMouseDown, onSingleNodeSelect))}
  </g>;

export default TreeView;
