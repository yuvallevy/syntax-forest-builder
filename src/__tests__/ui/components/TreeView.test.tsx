import { describe, expect, it } from 'vitest';
import { render } from '@testing-library/react';
import TreeView from '../../../ui/components/TreeView';
import { PositionedTree } from '../../../content/positioned/types';

describe('tree rendering', () => {
  const tree: PositionedTree = {
    sentence: 'Noun verbs very adverbly.',
    nodes: {
      'a': { label: 'S', position: { treeX: 53.625, treeY: -80 }, children: ['b', 'd'] },
      'b': { label: 'NP', position: { treeX: 18, treeY: -60 }, children: ['c'] },
      'c': { label: 'N', position: { treeX: 18, treeY: -2 }, slice: [0, 4] },
      'd': { label: 'VP', position: { treeX: 89.25, treeY: -60 }, children: ['e', 'f'] },
      'e': { label: 'V', position: { treeX: 57, treeY: -2 }, slice: [5, 10] },
      'f': { label: 'AdvP', position: { treeX: 121.5, treeY: -30 }, triangle: { treeX1: 72, treeX2: 104 }, slice: [11, 24] },
    },
    position: { plotX: 50, plotY: -32 },
    width: 104,
  };

  const treeWithUnlabeledNode: PositionedTree = {
    nodes: {
      'a': { label: 'CP', children: ['c', 'd'], position: { treeX: 37.375, treeY: -100 } },
      'b': { label: 'C\'', children: ['d', 'e'], position: { treeX: 59.25, treeY: -60 } },
      'c': { label: '', slice: [0, 3], position: { treeX: 15.5, treeY: -2 } },
      'd': { label: 'C', slice: [4, 7], position: { treeX: 47.5, treeY: -2 } },
      'e': { label: 'IP', slice: [8, 10], triangle: { treeX1: 64, treeX2: 78 }, position: { treeX: 71, treeY: -20 } },
    },
    sentence: 'Who was IP?',
    position: { plotX: 283, plotY: 238 },
    width: 86,
  };

  it('renders a tree with position-assigned nodes', () => {
    expect(render(
      <svg>  {/** Render inside an SVG tag to prevent "unrecognized tag" errors in console */}
        <TreeView treeId="tree1" tree={tree} selectedNodeIds={[]} />
      </svg>
    ).asFragment()).toMatchSnapshot();
  });

  it('renders a tree with position-assigned nodes where some are selected', () => {
    expect(render(
      <svg>
        <TreeView treeId="tree1" tree={tree} selectedNodeIds={['b', 'f']} />
      </svg>
    ).asFragment()).toMatchSnapshot();
  });

  it('renders dominated unlabeled nodes without text or gap', () => {
    expect(render(
      <svg>
        <TreeView treeId="tree1" tree={treeWithUnlabeledNode} selectedNodeIds={[]} />
      </svg>
    ).asFragment()).toMatchSnapshot();
  });
});
