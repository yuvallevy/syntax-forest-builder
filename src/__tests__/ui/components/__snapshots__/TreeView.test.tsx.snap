// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tree rendering > renders a tree with position-assigned nodes 1`] = `
<DocumentFragment>
  <svg>
      
    <g
      id="tree-tree1"
      style="transform: translate(50px, -32px);"
    >
      <g
        class="NodeCreationTriggerClickZone"
      >
        <rect
          class="NodeCreationTriggerClickZone--area"
          height="48"
          width="32"
          x="37.625"
          y="-148"
        />
        <circle
          class="NodeCreationTriggerClickZone--indicator"
          cx="53.625"
          cy="-120"
          r="8"
        />
        <line
          class="NodeCreationTriggerClickZone--indicator"
          x1="53.625"
          x2="53.625"
          y1="-120"
          y2="-100"
        />
      </g>
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="36.125"
          y="-98.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="53.625"
          y="-80"
        >
          S
        </text>
      </g>
      <line
        stroke="#000"
        x1="53.625"
        x2="18"
        y1="-80"
        y2="-80"
      />
      <line
        stroke="#000"
        x1="53.625"
        x2="89.25"
        y1="-80"
        y2="-80"
      />
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="0.5"
          y="-78.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="18"
          y="-60"
        >
          NP
        </text>
      </g>
      <line
        stroke="#000"
        x1="18"
        x2="18"
        y1="-60"
        y2="-22"
      />
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="0.5"
          y="-20.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="18"
          y="-2"
        >
          N
        </text>
      </g>
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="71.75"
          y="-78.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="89.25"
          y="-60"
        >
          VP
        </text>
      </g>
      <line
        stroke="#000"
        x1="89.25"
        x2="57"
        y1="-60"
        y2="-22"
      />
      <line
        stroke="#000"
        x1="89.25"
        x2="121.5"
        y1="-60"
        y2="-50"
      />
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="39.5"
          y="-20.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="57"
          y="-2"
        >
          V
        </text>
      </g>
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="104"
          y="-48.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="121.5"
          y="-30"
        >
          AdvP
        </text>
      </g>
      <path
        d="M121.5 -30 L72 -2  L104 -2 Z"
        fill="none"
        stroke="#000"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`tree rendering > renders a tree with position-assigned nodes where some are selected 1`] = `
<DocumentFragment>
  <svg>
    <g
      id="tree-tree1"
      style="transform: translate(50px, -32px);"
    >
      <g
        class="NodeCreationTriggerClickZone"
      >
        <rect
          class="NodeCreationTriggerClickZone--area"
          height="48"
          width="32"
          x="37.625"
          y="-148"
        />
        <circle
          class="NodeCreationTriggerClickZone--indicator"
          cx="53.625"
          cy="-120"
          r="8"
        />
        <line
          class="NodeCreationTriggerClickZone--indicator"
          x1="53.625"
          x2="53.625"
          y1="-120"
          y2="-100"
        />
      </g>
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="36.125"
          y="-98.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="53.625"
          y="-80"
        >
          S
        </text>
      </g>
      <line
        stroke="#000"
        x1="53.625"
        x2="18"
        y1="-80"
        y2="-80"
      />
      <line
        stroke="#000"
        x1="53.625"
        x2="89.25"
        y1="-80"
        y2="-80"
      />
      <g
        class="TreeView--node TreeView--node--selected"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="0.5"
          y="-78.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="18"
          y="-60"
        >
          NP
        </text>
      </g>
      <line
        stroke="#000"
        x1="18"
        x2="18"
        y1="-60"
        y2="-22"
      />
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="0.5"
          y="-20.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="18"
          y="-2"
        >
          N
        </text>
      </g>
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="71.75"
          y="-78.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="89.25"
          y="-60"
        >
          VP
        </text>
      </g>
      <line
        stroke="#000"
        x1="89.25"
        x2="57"
        y1="-60"
        y2="-22"
      />
      <line
        stroke="#000"
        x1="89.25"
        x2="121.5"
        y1="-60"
        y2="-50"
      />
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="39.5"
          y="-20.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="57"
          y="-2"
        >
          V
        </text>
      </g>
      <g
        class="TreeView--node TreeView--node--selected"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="104"
          y="-48.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="121.5"
          y="-30"
        >
          AdvP
        </text>
      </g>
      <path
        d="M121.5 -30 L72 -2  L104 -2 Z"
        fill="none"
        stroke="#000"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`tree rendering > renders dominated unlabeled nodes without text or gap 1`] = `
<DocumentFragment>
  <svg>
    <g
      id="tree-tree1"
      style="transform: translate(283px, 238px);"
    >
      <g
        class="NodeCreationTriggerClickZone"
      >
        <rect
          class="NodeCreationTriggerClickZone--area"
          height="48"
          width="32"
          x="21.375"
          y="-168"
        />
        <circle
          class="NodeCreationTriggerClickZone--indicator"
          cx="37.375"
          cy="-140"
          r="8"
        />
        <line
          class="NodeCreationTriggerClickZone--indicator"
          x1="37.375"
          x2="37.375"
          y1="-140"
          y2="-120"
        />
      </g>
      <g
        class="NodeCreationTriggerClickZone"
      >
        <rect
          class="NodeCreationTriggerClickZone--area"
          height="48"
          width="32"
          x="43.25"
          y="-128"
        />
        <circle
          class="NodeCreationTriggerClickZone--indicator"
          cx="59.25"
          cy="-100"
          r="8"
        />
        <line
          class="NodeCreationTriggerClickZone--indicator"
          x1="59.25"
          x2="59.25"
          y1="-100"
          y2="-80"
        />
      </g>
      <g
        class="NodeCreationTriggerClickZone"
      >
        <rect
          class="NodeCreationTriggerClickZone--area"
          height="48"
          width="32"
          x="32.3125"
          y="-168"
        />
        <circle
          class="NodeCreationTriggerClickZone--indicator"
          cx="48.3125"
          cy="-140"
          r="8"
        />
        <line
          class="NodeCreationTriggerClickZone--indicator"
          x1="48.3125"
          x2="37.375"
          y1="-140"
          y2="-120"
        />
        <line
          class="NodeCreationTriggerClickZone--indicator"
          x1="48.3125"
          x2="59.25"
          y1="-140"
          y2="-80"
        />
      </g>
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="19.875"
          y="-118.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="37.375"
          y="-100"
        >
          CP
        </text>
      </g>
      <line
        stroke="#000"
        x1="37.375"
        x2="15.5"
        y1="-100"
        y2="-2"
      />
      <line
        stroke="#000"
        x1="37.375"
        x2="47.5"
        y1="-100"
        y2="-22"
      />
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="41.75"
          y="-78.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="59.25"
          y="-60"
        >
          C'
        </text>
      </g>
      <line
        stroke="#000"
        x1="59.25"
        x2="47.5"
        y1="-60"
        y2="-22"
      />
      <line
        stroke="#000"
        x1="59.25"
        x2="71"
        y1="-60"
        y2="-40"
      />
      <g
        class="TreeView--node TreeView--node--empty-label"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="-2"
          y="-20.5"
        />
      </g>
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="30"
          y="-20.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="47.5"
          y="-2"
        >
          C
        </text>
      </g>
      <g
        class="TreeView--node"
      >
        <rect
          height="20"
          rx="3"
          ry="3"
          width="35"
          x="53.5"
          y="-38.5"
        />
        <text
          dominant-baseline="text-after-edge"
          fill="#000"
          text-anchor="middle"
          x="71"
          y="-20"
        >
          IP
        </text>
      </g>
      <path
        d="M71 -20 L64 -2  L78 -2 Z"
        fill="none"
        stroke="#000"
      />
    </g>
  </svg>
</DocumentFragment>
`;
